glom/core.py

```diff
@@ -100,10 +115,126 @@ class GlomError(Exception):
     (e.g., ``len``, ``sum``, any ``lambda``) will not be wrapped in a
     GlomError.
     """
-    pass
+    @classmethod
+    def wrap(cls, exc):
+        # TODO: need to test this against a wide array of exception types
+        # this approach to wrapping errors works for exceptions
+        # defined in pure-python as well as C
+        exc_type = type(exc)
+        bases = (GlomError,) if issubclass(GlomError, exc_type) else (exc_type, GlomError)
+        exc_wrapper_type = type("GlomError.wrap({})".format(exc_type.__name__), bases, {})
+        try:
+            return exc_wrapper_type(*exc.args)
+        except Exception:  # maybe exception can't be re-created
+            return exc
```

Comment:
this approach looks like it'll work. i hope it's not adding much overhead, but it's the exception path. then again, exception path gets hit pretty often for folks using defaults and stuff. 

eh, we'll benchmark it when we benchmark it. i think this is probably even a little safer than we need to be. if we end up optimizing this away later, once we know how corner casey this is, so be it.